control/controller.go

```diff
@@ -540,40 +549,40 @@ func (q *Query) pump(exec flux.Query) {
 			// done, nobody is going to read these values so we need
 			// to avoid blocking.
 			select {
-			case <-q.currentCtx.Done():
+			case <-done:
 			case q.results <- res:
 			}
-		case <-done:
+		case <-signalCh:
 			// Signal to the underlying executor that the query
 			// has been canceled. Usually, the signal on the context
 			// is likely enough, but this explicitly signals just in case.
 			exec.Cancel()
 
 			// Set the done channel to nil so we don't do this again
 			// and we continue to drain the results.
-			done = nil
+			signalCh = nil
 		}
 	}
 }
 
 // tryCompile attempts to transition the query into the Compiling state.
-func (q *Query) tryCompile() bool {
+func (q *Query) tryCompile() (context.Context, bool) {
 	q.stateMu.Lock()
```

Comment:
Should `stateMu` just be locked inside the `transitionTo` function instead?

Reply:
Probably. It wasn't originally because `transitionTo` was sometimes called within a function that locked the mutex, but it no longer appears to do that. It's possible this should be modified to return an `error` too.

Reply:
When I went to modify this, I found that `transitionTo` invokes itself so it would be hard to put the lock in there without doing a larger change than I'm comfortable with so I'm going to leave this as-is.